man/man1/mkstr.1
.\" @(#)mkstr.1 1.1 92/07/30 SMI; from UCB 4.1
.TH MKSTR 1 "9 September 1987"
.SH NAME
mkstr \- create an error message file by massaging C source files
.SH SYNOPSIS
.B mkstr
[
.B \-
]
.I messagefile
.I prefix
.IR filename .\|.\|.
.SH DESCRIPTION
.IX  directory  make  ""  "make \(em \fLmkdir\fP"
.IX  "mkstr command"  ""  "\fLmkstr\fP \(em create C error messages"
.IX  "programming tools"  mkstr  ""  "\fLmkstr\fP \(em create C error messages"
.IX  "C programming language"  mkstr  ""  "\fLmkstr\fP \(em create C error messages"
.IX  languages  mkstr  ""  "\fLmkstr\fP \(em create C error messages"
.B mkstr
creates files of error messages.
You can use
.B mkstr
to make programs with large numbers of error
diagnostics much smaller, and to reduce
system overhead in running the
program \(em as the error messages
do not have to be constantly swapped in
and out.
.LP
.B mkstr
processes each of the specified
.IR filename s,
placing a massaged version of the input file in a file with
a name consisting of the specified
.I prefix
and the original source file name.
A typical example of using
.B mkstr
would be:
.IP
.B mkstr pistrings processed *.c
.LP
This command would cause all the error messages from the C source
files in the current directory to be placed in the file
.B pistrings
and processed copies of the source for these files to be placed in
files whose names are prefixed with
.IR processed .
.LP
To process the error messages in the source to the message file,
.B mkstr
keys on the string `\fBerror("\fR'
in the input stream.
Each time it occurs, the C string starting at the `\fB"\fR'
is placed in the message file followed by a null character and a
.SM NEWLINE
character; the null character terminates
the message so it can be easily used
when retrieved, the
.SM NEWLINE
character makes it possible to sensibly
.B cat
the error message file to see its contents.
The massaged copy of the input file then contains a
.B lseek
pointer into the file which can be used
to retrieve the message, that is:
.RS
.IP
.nf
.ft B
char efilname[\|] = "/usr/lib/pi_strings";
int efil = \-1;
.sp
error(a1, a2, a3, a4)
{
.RS
.IP
.ft B
char
buf[256];
if (efil < 0) {
.RS
.RS
.IP
.ft B
efil = open(efilname, 0);
if (efil < 0) {
.RE
.RE
.RE
.B oops:
.RS
.RS
.RS
.IP
.B perror (efilname);
.B exit (\&1);
.RE
.B }
.RE
.ft B
}
if (lseek(efil, (long) a1, 0) |\|| read(efil, buf, 256) <= 0)
.RS
.B goto oops;
.RE
.B printf(buf, a2, a3, a4);
.RE
.B }
.fi
.ft R
.RE
.SH OPTIONS
.TP 3
.B \-
Place error messages at the end of the specified
message file for recompiling part of a large
\fBmkstr\fPed
program.
.SH SEE\ ALSO
.BR xstr (1)


ucb/mkstr.c
#ifndef lint
static	char sccsid[] = "@(#)mkstr.c 1.1 92/07/30 SMI"; /* from UCB 4.1 10/01/80 */
#endif
#include <stdio.h>

#define	ungetchar(c)	ungetc(c, stdin)

long	ftell();
char	*calloc();
/*
 * mkstr - create a string error message file by massaging C source
 *
 * Bill Joy UCB August 1977
 *
 * Modified March 1978 to hash old messages to be able to recompile
 * without addding messages to the message file (usually)
 *
 * Based on an earlier program conceived by Bill Joy and Chuck Haley
 *
 * Program to create a string error message file
 * from a group of C programs.  Arguments are the name
 * of the file where the strings are to be placed, the
 * prefix of the new files where the processed source text
 * is to be placed, and the files to be processed.
 *
 * The program looks for 'error("' in the source stream.
 * Whenever it finds this, the following characters from the '"'
 * to a '"' are replaced by 'seekpt' where seekpt is a
 * pointer into the error message file.
 * If the '(' is not immediately followed by a '"' no change occurs.
 *
 * The optional '-' causes strings to be added at the end of the
 * existing error message file for recompilation of single routines.
 */


FILE	*mesgread, *mesgwrite;
char	*progname;
char	usagestr[] =	"usage: %s [ - ] mesgfile prefix file ...\n";
char	name[100], *np;

main(argc, argv)
	int argc;
	char *argv[];
{
	char addon = 0;

	argc--, progname = *argv++;
	if (argc > 1 && argv[0][0] == '-')
		addon++, argc--, argv++;
	if (argc < 3)
		fprintf(stderr, usagestr, progname), exit(1);
	mesgwrite = fopen(argv[0], addon ? "a" : "w");
	if (mesgwrite == NULL)
		perror(argv[0]), exit(1);
	mesgread = fopen(argv[0], "r");
	if (mesgread == NULL)
		perror(argv[0]), exit(1);
	inithash();
	argc--, argv++;
	strcpy(name, argv[0]);
	np = name + strlen(name);
	argc--, argv++;
	do {
		strcpy(np, argv[0]);
		if (freopen(name, "w", stdout) == NULL)
			perror(name), exit(1);
		if (freopen(argv[0], "r", stdin) == NULL)
			perror(argv[0]), exit(1);
		process();
		argc--, argv++;
	} while (argc > 0);
	exit(0);
	/* NOTREACHED */
}

process()
{
	register char *cp;
	register c;

	for (;;) {
		c = getchar();
		if (c == EOF)
			return;
		if (c != 'e') {
			putchar(c);
			continue;
		}
		if (match("error(")) {
			printf("error(");
			c = getchar();
			if (c != '"')
				putchar(c);
			else
				copystr();
		}
	}
}

match(ocp)
	char *ocp;
{
	register char *cp;
	register c;

	for (cp = ocp + 1; *cp; cp++) {
		c = getchar();
		if (c != *cp) {
			while (ocp < cp)
				putchar(*ocp++);
			ungetchar(c);
			return (0);
		}
	}
	return (1);
}

copystr()
{
	register c, ch;
	char buf[512];
	register char *cp = buf;

	for (;;) {
		c = getchar();
		if (c == EOF)
			break;
		switch (c) {

		case '"':
			*cp++ = 0;
			goto out;
		case '\\':
			c = getchar();
			switch (c) {

			case 'b':
				c = '\b';
				break;
			case 't':
				c = '\t';
				break;
			case 'r':
				c = '\r';
				break;
			case 'n':
				c = '\n';
				break;
			case '\n':
				continue;
			case 'f':
				c = '\f';
				break;
			case '0':
				c = 0;
				break;
			case '\\':
				break;
			default:
				if (!octdigit(c))
					break;
				c -= '0';
				ch = getchar();
				if (!octdigit(ch))
					break;
				c <<= 7, c += ch - '0';
				ch = getchar();
				if (!octdigit(ch))
					break;
				c <<= 3, c+= ch - '0', ch = -1;
				break;
			}
		}
		*cp++ = c;
	}
out:
	*cp = 0;
	printf("%d", hashit(buf, 1, NULL));
}

octdigit(c)
	char c;
{

	return (c >= '0' && c <= '7');
}

inithash()
{
	char buf[512];
	int mesgpt = 0;

	rewind(mesgread);
	while (fgetNUL(buf, sizeof buf, mesgread) != NULL) {
		hashit(buf, 0, mesgpt);
		mesgpt += strlen(buf) + 2;
	}
}

#define	NBUCKETS	511

struct	hash {
	long	hval;
	unsigned hpt;
	struct	hash *hnext;
} *bucket[NBUCKETS];

hashit(str, really, fakept)
	char *str;
	char really;
	unsigned fakept;
{
	int i;
	register struct hash *hp;
	char buf[512];
	long hashval = 0;
	register char *cp;

	if (really)
		fflush(mesgwrite);
	for (cp = str; *cp;)
		hashval = (hashval << 1) + *cp++;
	i = hashval % NBUCKETS;
	if (i < 0)
		i += NBUCKETS;
	if (really != 0)
		for (hp = bucket[i]; hp != 0; hp = hp->hnext)
		if (hp->hval == hashval) {
			fseek(mesgread, (long) hp->hpt, 0);
			fgetNUL(buf, sizeof buf, mesgread);
/*
			fprintf(stderr, "Got (from %d) %s\n", hp->hpt, buf);
*/
			if (strcmp(buf, str) == 0)
				break;
		}
	if (!really || hp == 0) {
		hp = (struct hash *) calloc(1, sizeof *hp);
		hp->hnext = bucket[i];
		hp->hval = hashval;
		hp->hpt = really ? ftell(mesgwrite) : fakept;
		if (really) {
			fwrite(str, sizeof (char), strlen(str) + 1, mesgwrite);
			fwrite("\n", sizeof (char), 1, mesgwrite);
		}
		bucket[i] = hp;
	}
/*
	fprintf(stderr, "%s hashed to %ld at %d\n", str, hp->hval, hp->hpt);
*/
	return (hp->hpt);
}

#include <sys/types.h>
#include <sys/stat.h>

fgetNUL(obuf, rmdr, file)
	char *obuf;
	register int rmdr;
	FILE *file;
{
	register c;
	register char *buf = obuf;

	while (--rmdr > 0 && (c = getc(file)) != 0 && c != EOF)
		*buf++ = c;
	*buf++ = 0;
	getc(file);
	return ((feof(file) || ferror(file)) ? NULL : 1);
}
